File: pnpm11/installing/env-installer/src/resolvePackageManagerIntegrities.ts
import { parseRegistryQualifiedVersion, refToRelative, removeSuffix } from '@pnpm/deps.path'
import { PnpmError } from '@pnpm/error'
import { convertToLockfileFile, createEnvLockfile, readEnvLockfile } from '@pnpm/lockfile.fs'
import { pruneSharedLockfile } from '@pnpm/lockfile.pruner'
import type { EnvLockfile, LockfileObject } from '@pnpm/lockfile.types'
import type { StoreController } from '@pnpm/store.controller'
import type { DepPath, ProjectId, RegistriesByScope } from '@pnpm/types'
import semver from 'semver'

import { convertToLockfileEnvObject } from './pruneEnvLockfile.js'
import { resolveManifestDependencies } from './resolveManifestDependencies.js'
import { writeVerifiedEnvLockfile } from './writeVerifiedEnvLockfile.js'

const PACKAGE_MANAGER_DEPS_WITH_EXE = ['pnpm', '@pnpm/exe'] as const
const PACKAGE_MANAGER_DEPS_PNPM_ONLY = ['pnpm'] as const
const PNPM_EXE_INTRODUCED = '6.17.1'

export interface ResolvePackageManagerIntegritiesOpts {
  envLockfile?: EnvLockfile
  registriesByScope: RegistriesByScope
  rootDir: string
  storeController: StoreController
  storeDir: string
  /**
   * Whether to read from and write to the env lockfile file on disk.
   * When false, resolution happens purely in memory; callers can still use
   * the returned `EnvLockfile` to perform installs without persisting the
   * resolved pnpm integrity info. Defaults to true.
   */
  save?: boolean
  /**
   * The specifier to record for each entry, when it differs from the version
   * being resolved. A range pin names no exact version, so its entries resolve
   * to a concrete version while still recording the range the project asked
   * for. Defaults to the resolved spec, which is what an exact pin wants.
   */
  specifier?: string
  /**
   * Refuse to record entries that are missing or out of date, failing the
   * command instead. Only meaningful together with `save`: an in-memory
   * resolution changes no lockfile, so nothing can fall out of sync with it.
   */
  frozenLockfile?: boolean
}

/**
 * Checks if the wanted pnpm version integrities are already fully resolved in the env lockfile.
 *
 * `specifier` is what the project asked for, which a range pin records
 * alongside the version it resolved to. Pass it to also require that the entry
 * records the pin the project asks for now: one written under a pin that has
 * since changed still has to be rewritten, even though its version stands.
 * Omit it to accept the entry on its version alone.
 */
export function isPackageManagerResolved (
  envLockfile: EnvLockfile | undefined,
  pnpmVersion: string,
  specifier?: string
): boolean {
  if (!envLockfile) return false

  const pmDeps = envLockfile.importers['.'].packageManagerDependencies
  if (pmDeps == null) return false
  const wantedDeps = packageManagerDeps(pnpmVersion)
  return Object.keys(pmDeps).length === wantedDeps.length &&
    wantedDeps.every((name) =>
      pmDeps[name]?.version === pnpmVersion &&
      (specifier == null || pmDeps[name]?.specifier === specifier)
    )
}

/**
 * Whether the env lockfile pins the package manager the manifest asks for,
 * even when it records more packages than this pnpm installs it from.
 *
 * A pnpm below 11.20.0 pins `@pnpm/exe` beside `pnpm` for a v12 version,
 * because that is the set its own major is installed from. Such an entry pins
 * the wanted version through the same integrity and cannot change which pnpm
 * runs, so a frozen lockfile accepts it instead of failing a project whose
 * lockfile a teammate's older pnpm last wrote. An entry pinning any other
 * version, or one the lockfile carries no package to install from, is a
 * lockfile that disagrees with the manifest, which is what the flag is for,
 * and a writable install still rewrites the block to the packages this pnpm
 * installs from.
 */
export function pinsWantedPackageManager (
  envLockfile: EnvLockfile | undefined,
  pnpmVersion: string
): boolean {
  if (!envLockfile) return false

  const pmDeps = envLockfile.importers['.'].packageManagerDependencies
  if (pmDeps == null) return false
  return packageManagerDeps(pnpmVersion).every((name) => pmDeps[name] != null) &&
    Object.entries(pmDeps).every(([name, dep]) =>
      dep.version === pnpmVersion && isRecordedForInstall(envLockfile, name, dep.version)
    )
}

/** Whether the lockfile carries the records the bootstrap installs `name@version` from. */
function isRecordedForInstall (envLockfile: EnvLockfile, name: string, version: string): boolean {
  const depPath = refToRelative(version, name)
  return depPath != null &&
    envLockfile.packages[removeSuffix(depPath)] != null &&
    envLockfile.snapshots[depPath] != null
}

/**
 * The packages the env lockfile pins for `pnpmVersion`.
 *
 * `>=6.17.1 <12` publishes the JS `pnpm` and the native `@pnpm/exe` as two
 * packages, and both are pinned, because the pin is shared and teammates may
 * run either one. Every other version publishes `pnpm` alone: as the JS CLI
 * below 6.17.1, and as the native executable itself from 12.
 */
function packageManagerDeps (pnpmVersion: string): readonly string[] {
  const parsed = semver.parse(pnpmVersion, { loose: true })
  if (parsed == null) return PACKAGE_MANAGER_DEPS_WITH_EXE
  if (parsed.major >= 12) return PACKAGE_MANAGER_DEPS_PNPM_ONLY
  // Prereleases of a version that has `@pnpm/exe` ship it too, so compare on
  // the release triple alone.
  return semver.gte(`${parsed.major}.${parsed.minor}.${parsed.patch}`, PNPM_EXE_INTRODUCED)
    ? PACKAGE_MANAGER_DEPS_WITH_EXE
    : PACKAGE_MANAGER_DEPS_PNPM_ONLY
}

/**
 * Resolves integrity checksums for the pnpm packages of the wanted version
 * (see {@link packageManagerDeps}) and their dependencies by calling
 * resolveManifestDependencies. When `opts.save` is true (the default) the
 * results are written to the `packageManagerDependencies` section of
 * `pnpm-lock.yaml`; when false, resolution happens purely in memory and the
 * returned `EnvLockfile` is never persisted to disk. Under
 * `opts.frozenLockfile` a write the lockfile still needs is an error instead.
 */
export async function resolvePackageManagerIntegrities (
  pnpmVersion: string,
  opts: ResolvePackageManagerIntegritiesOpts
): Promise<EnvLockfile> {
  const save = opts.save ?? true
  const envLockfile = opts.envLockfile ?? (save ? await readEnvLockfile(opts.rootDir) : undefined) ?? createEnvLockfile()

  // A frozen lockfile is not rewritten to refresh a stale specifier: refusing
  // the entry over one would fail the install rather than update it.
  if (isPackageManagerResolved(envLockfile, pnpmVersion, opts.frozenLockfile ? undefined : opts.specifier)) {
    return envLockfile
  }

  if (save && opts.frozenLockfile) {
    if (pinsWantedPackageManager(envLockfile, pnpmVersion)) {
      return envLockfile
    }
    throw new PnpmError('FROZEN_LOCKFILE_WITH_OUTDATED_LOCKFILE', 'Cannot update packageManagerDependencies with "frozen-lockfile" because the lockfile is not up to date')
  }

  const lockfile = await resolveWantedPnpmPackages(pnpmVersion, opts)
  stripRegistryTarballUrls(lockfile)

  if (lockfile.packages) {
    // Build packageManagerDependencies from the resolved lockfile importers
    const importer = lockfile.importers['.' as ProjectId]
    const packageManagerDependencies: Record<string, { specifier: string, version: string }> = {}
    for (const [name, version] of Object.entries(importer.dependencies ?? {})) {
      packageManagerDependencies[name] = {
        specifier: importer.specifiers[name],
        version,
      }
    }
    envLockfile.importers['.'].packageManagerDependencies = packageManagerDependencies

    // Merge new packages into the env lockfile object, then prune stale entries
    const merged = convertToLockfileEnvObject(envLockfile)
    for (const [depPath, pkg] of Object.entries(lockfile.packages)) {
      merged.packages![depPath as DepPath] = pkg
    }
    const pruned = pruneSharedLockfile(merged)
    const prunedFile = convertToLockfileFile(pruned)
    envLockfile.packages = prunedFile.packages ?? {}
    envLockfile.snapshots = prunedFile.snapshots ?? {}

    if (save) {
      await writeVerifiedEnvLockfile(opts.rootDir, envLockfile)
    }
  }
  return envLockfile
}

/**
 * Rewrites registry tarball resolutions to integrity-only form, dropping
 * tarball URLs that a registry advertises on a host other than its own —
 * load-balanced proxies and Artifactory-style mirrors do this, see
 * https://github.com/pnpm/pnpm/issues/13619. The package-manager bootstrap
 * never fetches a URL recorded in the lockfile: the download URL is always
 * derived from the trusted bootstrap registries at install time, so a
 * repository-provided entry cannot steer the download. Dropping the URL here
 * keeps freshly resolved entries in exactly the integrity-only shape the
 * bootstrap validation accepts.
 */
function stripRegistryTarballUrls (lockfile: LockfileObject): void {
  for (const pkg of Object.values(lockfile.packages ?? {})) {
    const resolution = pkg.resolution
    if (
      resolution == null ||
      !('integrity' in resolution) || !resolution.integrity ||
      !('tarball' in resolution) || typeof resolution.tarball !== 'string' ||
      resolution.tarball.startsWith('file:') ||
      ('gitHosted' in resolution && resolution.gitHosted === true) ||
      ('path' in resolution && resolution.path != null)
    ) {
      continue
    }
    pkg.resolution = { integrity: resolution.integrity }
  }
}

/**
 * Resolves the pnpm packages wanted by `spec`, which may also be a range or a
 * dist-tag. A spec that is not an exact version is resolved through `pnpm`
 * alone first, because which packages are wanted (see
 * {@link packageManagerDeps}) is only known once the version is known.
 */
async function resolveWantedPnpmPackages (
  spec: string,
  opts: ResolvePackageManagerIntegritiesOpts
): Promise<LockfileObject> {
  const resolveOpts = {
    dir: opts.rootDir,
    registriesByScope: opts.registriesByScope,
    storeController: opts.storeController,
    storeDir: opts.storeDir,
  }
  if (semver.valid(spec) != null) {
    return resolveManifestDependencies({ dependencies: wantedDependencies(spec, opts.specifier ?? spec) }, resolveOpts)
  }
  const lockfile = await resolveManifestDependencies({ dependencies: { pnpm: spec } }, resolveOpts)
  const ref = lockfile.importers['.' as ProjectId]?.dependencies?.['pnpm']
  // A pnpm resolved from a named registry is referenced as `<registry>:<version>`.
  const resolvedVersion = ref == null ? undefined : parseRegistryQualifiedVersion(ref)?.version ?? ref
  if (resolvedVersion == null || !packageManagerDeps(resolvedVersion).includes('@pnpm/exe')) {
    return lockfile
  }
  return resolveManifestDependencies({ dependencies: wantedDependencies(resolvedVersion, spec) }, resolveOpts)
}

/** The dependency map to resolve `spec` through, for a pnpm of `pnpmVersion`. */
function wantedDependencies (pnpmVersion: string, spec: string): Record<string, string> {
  return Object.fromEntries(packageManagerDeps(pnpmVersion).map((name) => [name, spec]))
}

File: pnpm/crates/env-installer/src/resolve_package_manager_integrities.rs
use crate::{
    ConfigDepError,
    manifest_lockfile::{package_metadata, read_dependency_map},
    options::ConfigDepsInstallOptions,
    prune::prune_env_lockfile,
    resolve_optional_subdeps::resolution_has_integrity,
    verify_env_lockfile::{verify_env_lockfile, write_verified_env_lockfile},
};
use pnpm_lockfile::{
    EnvLockfile, LockfileResolution, PackageKey, PkgName, PkgVerPeer, RegistryResolution,
    SnapshotDepRef, SnapshotEntry, SpecifierAndResolution, TarballResolution,
};
use pnpm_resolving_resolver_base::{ResolveOptions, ResolveResult, Resolver, WantedDependency};
use std::{collections::HashMap, path::PathBuf};

const PACKAGE_MANAGER_DEPS_WITH_EXE: [&str; 2] = ["pnpm", "@pnpm/exe"];
const PACKAGE_MANAGER_DEPS_PNPM_ONLY: [&str; 1] = ["pnpm"];
const PNPM_EXE_INTRODUCED: (u64, u64, u64) = (6, 17, 1);

/// Resolve the closure of `package_manager_deps` — the packages a package
/// manager is installed from — at `version`, record it in the env lockfile
/// at `opts.root_dir`, and return that lockfile.
///
/// `force_resync` skips the recorded-entries fast path, so entries that look
/// up to date but are invalid (e.g. resolutions carrying tarball URLs
/// written by an earlier pnpm) are discarded and re-resolved. Such entries
/// already record the version the manifest pins — only pnpm's own reading of
/// them is at stake — so under `opts.frozen_lockfile` the re-resolution
/// happens in memory and the lockfile is left untouched, rather than failing
/// a command the manifest and the lockfile agree on. Entries that do not
/// record the pinned version are the case `--frozen-lockfile` exists for and
/// still fail.
pub async fn resolve_package_manager_integrities(
    package_manager_deps: &[&str],
    wanted_specifier: &str,
    version: &str,
    resolver: &dyn Resolver,
    opts: &ConfigDepsInstallOptions<'_>,
    force_resync: bool,
) -> Result<EnvLockfile, ConfigDepError> {
    let mut env_lockfile = EnvLockfile::read(opts.root_dir)
        .map_err(ConfigDepError::ReadLockfile)?
        .unwrap_or_else(EnvLockfile::create);
    if !force_resync
        && is_package_manager_resolved_with_deps(
            &env_lockfile,
            wanted_specifier,
            version,
            package_manager_deps,
        )
    {
        return Ok(env_lockfile);
    }
    let repair_in_memory = force_resync && opts.frozen_lockfile;
    if opts.frozen_lockfile && !force_resync {
        if pins_wanted_package_manager(&env_lockfile, version, package_manager_deps) {
            return Ok(env_lockfile);
        }
        return Err(ConfigDepError::FrozenLockfileOutdated {
            message: r#"Cannot update packageManagerDependencies with "frozen-lockfile" because the lockfile is not up to date"#.to_string(),
        });
    }

    let mut package_manager_dependencies = std::collections::BTreeMap::new();
    let mut resolved = Vec::new();
    for name in package_manager_deps {
        let package = resolve_dep(name, version, false, resolver, opts).await?;
        package_manager_dependencies.insert(
            (*name).to_string(),
            SpecifierAndResolution {
                specifier: wanted_specifier.to_string(),
                version: package.version.clone(),
            },
        );
        resolved.push(package);
    }

    env_lockfile.root_importer_mut().package_manager_dependencies =
        Some(package_manager_dependencies);

    let mut seen = std::collections::HashSet::new();
    while let Some(package) = resolved.pop() {
        if !seen.insert(package.key.clone()) {
            if !package.optional
                && let Some(snapshot) = env_lockfile.snapshots.get_mut(&package.key)
            {
                snapshot.optional = false;
            }
            continue;
        }
        let registry = opts.pick_registry(&package.name);
        let mut metadata =
            package_metadata(&package.name, &package.version, &package.result, registry, false)
                .map_err(ConfigDepError::LockfileForm)?;
        metadata.resolution = strip_registry_tarball_url(metadata.resolution);
        env_lockfile.packages.insert(package.key.clone(), metadata);

        let manifest = package.result.manifest.as_deref();
        let mut dependencies = HashMap::new();
        for (alias, specifier) in read_dependency_map(manifest, "dependencies") {
            let child = resolve_dep(&alias, &specifier, false, resolver, opts).await?;
            dependencies.insert(snapshot_dep_name(&alias)?, child.snapshot_ref(&alias)?);
            resolved.push(child);
        }

        let mut optional_dependencies = HashMap::new();
        for (alias, specifier) in read_dependency_map(manifest, "optionalDependencies") {
            let child = resolve_dep(&alias, &specifier, true, resolver, opts).await?;
            optional_dependencies.insert(snapshot_dep_name(&alias)?, child.snapshot_ref(&alias)?);
            resolved.push(child);
        }

        env_lockfile.snapshots.insert(
            package.key,
            SnapshotEntry {
                dependencies: (!dependencies.is_empty()).then_some(dependencies),
                optional_dependencies: (!optional_dependencies.is_empty())
                    .then_some(optional_dependencies),
                optional: package.optional,
                ..SnapshotEntry::default()
            },
        );
    }

    prune_env_lockfile(&mut env_lockfile);
    if repair_in_memory {
        verify_env_lockfile(&env_lockfile)?;
    } else {
        write_verified_env_lockfile(&env_lockfile, opts.root_dir)?;
    }
    Ok(env_lockfile)
}

/// Rewrite a registry tarball resolution to integrity-only form, dropping
/// tarball URLs a registry advertises on a host other than its own —
/// load-balanced proxies and Artifactory-style mirrors do this, see
/// <https://github.com/pnpm/pnpm/issues/13619>. The package-manager
/// bootstrap never fetches a URL recorded in the lockfile: the download URL
/// is always derived from the trusted bootstrap registries at install time,
/// so a repository-provided entry cannot steer the download. Dropping the
/// URL here keeps freshly resolved entries in exactly the integrity-only
/// shape the bootstrap validation accepts.
fn strip_registry_tarball_url(resolution: LockfileResolution) -> LockfileResolution {
    match resolution {
        LockfileResolution::Tarball(TarballResolution {
            tarball,
            integrity: Some(integrity),
            revision: None,
            git_hosted: None | Some(false),
            path: None,
        }) if !tarball.starts_with("file:") => {
            LockfileResolution::Registry(RegistryResolution { integrity, revision: None })
        }
        other => other,
    }
}

#[must_use]
pub fn is_package_manager_resolved(
    env_lockfile: &EnvLockfile,
    wanted_specifier: &str,
    pnpm_version: &str,
) -> bool {
    is_package_manager_resolved_with_deps(
        env_lockfile,
        wanted_specifier,
        pnpm_version,
        pnpm_engine_packages(pnpm_version),
    )
}

/// Whether the env lockfile already records what this pnpm would write for
/// `pnpm_version`: the pinned packages under the specifier they were pinned
/// from, and nothing besides them.
fn is_package_manager_resolved_with_deps(
    env_lockfile: &EnvLockfile,
    wanted_specifier: &str,
    pnpm_version: &str,
    package_manager_deps: &[&str],
) -> bool {
    recorded_package_manager_deps(env_lockfile).is_some_and(|pm_deps| {
        pm_deps.len() == package_manager_deps.len()
            && pm_deps.values().all(|dep| dep.specifier == wanted_specifier)
    }) && pins_wanted_package_manager(env_lockfile, pnpm_version, package_manager_deps)
}

/// Whether the env lockfile pins the package manager the manifest asks for,
/// even when it records more packages than this pnpm installs it from.
///
/// A pnpm below 11.20.0 pins `@pnpm/exe` beside `pnpm` for a v12 version,
/// because that is the set its own major is installed from. Such an entry
/// pins the wanted version through the same integrity and cannot change
/// which pnpm runs, so a frozen lockfile accepts it instead of failing a
/// project whose lockfile a teammate's older pnpm last wrote. An entry
/// pinning any other version, or one the lockfile carries no package to
/// install from, is a lockfile that disagrees with the manifest, which is
/// what the flag is for, and a writable install still rewrites the block to
/// the packages this pnpm installs from.
fn pins_wanted_package_manager(
    env_lockfile: &EnvLockfile,
    pnpm_version: &str,
    package_manager_deps: &[&str],
) -> bool {
    let Some(pm_deps) = recorded_package_manager_deps(env_lockfile) else {
        return false;
    };
    package_manager_deps.iter().all(|name| pm_deps.contains_key(*name))
        && pm_deps.iter().all(|(name, dep)| {
            dep.version == pnpm_version
                && package_manager_entry_exists(env_lockfile, name, &dep.version)
        })
}

fn recorded_package_manager_deps(
    env_lockfile: &EnvLockfile,
) -> Option<&std::collections::BTreeMap<String, SpecifierAndResolution>> {
    env_lockfile
        .importers
        .get(EnvLockfile::ROOT_IMPORTER_KEY)
        .and_then(|importer| importer.package_manager_dependencies.as_ref())
}

/// The packages the env lockfile pins for `pnpm_version`.
///
/// `>=6.17.1 <12` publishes the JS `pnpm` and the native `@pnpm/exe` as two
/// packages, and both are pinned, because the pin is shared and teammates
/// may run either one. Every other version publishes `pnpm` alone: as the JS
/// CLI below 6.17.1, and as the native executable itself from 12.
#[must_use]
pub fn pnpm_engine_packages(pnpm_version: &str) -> &'static [&'static str] {
    let Some(version) = node_semver::Version::parse(pnpm_version).ok() else {
        return &PACKAGE_MANAGER_DEPS_WITH_EXE;
    };
    if version.major >= 12 {
        return &PACKAGE_MANAGER_DEPS_PNPM_ONLY;
    }
    if (version.major, version.minor, version.patch) >= PNPM_EXE_INTRODUCED {
        &PACKAGE_MANAGER_DEPS_WITH_EXE
    } else {
        &PACKAGE_MANAGER_DEPS_PNPM_ONLY
    }
}

fn package_manager_entry_exists(env_lockfile: &EnvLockfile, name: &str, version: &str) -> bool {
    let Ok(key) = format!("{name}@{version}").parse::<PackageKey>() else {
        return false;
    };
    env_lockfile.packages.contains_key(&key) && env_lockfile.snapshots.contains_key(&key)
}

struct EnvPackage {
    name: String,
    version: String,
    key: PackageKey,
    optional: bool,
    result: ResolveResult,
}

impl EnvPackage {
    fn snapshot_ref(&self, alias: &str) -> Result<SnapshotDepRef, ConfigDepError> {
        if alias == self.name {
            let ver_peer =
                self.version.parse::<PkgVerPeer>().map_err(|_| ConfigDepError::BadConfigDep {
                    message: format!(
                        "Resolved package manager dependency version {} is not valid",
                        self.version,
                    ),
                })?;
            Ok(SnapshotDepRef::Plain(ver_peer))
        } else {
            format!("{}@{}", self.name, self.version).parse().map(SnapshotDepRef::Alias).map_err(
                |_| ConfigDepError::BadConfigDep {
                    message: format!(
                        "Resolved package manager dependency {}@{} has an unparsable alias reference",
                        self.name, self.version,
                    ),
                },
            )
        }
    }
}

async fn resolve_dep(
    alias: &str,
    specifier: &str,
    optional: bool,
    resolver: &dyn Resolver,
    opts: &ConfigDepsInstallOptions<'_>,
) -> Result<EnvPackage, ConfigDepError> {
    let wanted = WantedDependency {
        alias: Some(alias.to_string()),
        bare_specifier: Some(specifier.to_string()),
        optional: optional.then_some(true),
        ..WantedDependency::default()
    };
    let resolve_opts = ResolveOptions {
        project_dir: PathBuf::from(opts.root_dir),
        lockfile_dir: PathBuf::from(opts.root_dir),
        ..ResolveOptions::default()
    };
    let result = resolver
        .resolve(&wanted, &resolve_opts)
        .await
        .map_err(|error| ConfigDepError::Resolve { spec: format!("{alias}@{specifier}"), error })?
        .ok_or_else(|| no_integrity(alias, specifier))?;
    if !resolution_has_integrity(&result.resolution) {
        return Err(no_integrity(alias, specifier));
    }
    let name_ver = result.name_ver.as_ref().ok_or_else(|| no_integrity(alias, specifier))?;
    let name = name_ver.name.to_string();
    let version = name_ver.suffix.to_string();
    let key = format!("{name}@{version}").parse::<PackageKey>().map_err(|_| {
        ConfigDepError::BadConfigDep {
            message: format!(
                "Resolved package manager dependency {name}@{version} has an unparsable lockfile key",
            ),
        }
    })?;
    Ok(EnvPackage { name, version, key, optional, result })
}

fn no_integrity(alias: &str, specifier: &str) -> ConfigDepError {
    ConfigDepError::BadConfigDep {
        message: format!(
            "Cannot resolve {alias}@{specifier} as a package manager dependency because it has no integrity",
        ),
    }
}

fn snapshot_dep_name(alias: &str) -> Result<PkgName, ConfigDepError> {
    alias.parse().map_err(|_| ConfigDepError::BadConfigDep {
        message: format!("Resolved package manager dependency name {alias} is invalid"),
    })
}
